### PR TITLE
CE-95: Retrofitting error collection and adding tests

### PR DESCRIPTION
**Pull request by ialarmedalien**
Collect errors instead of emitting them one at a time.

### CODE REVIEW

File: src/us/kbase/workspace/database/provenance/EventDate.java [hunk 1]
```diff
@@ -62,34 +65,7 @@ public boolean equals(Object obj) {
...
-	 * @return the new {@link EventDate}.
-	 */
-	public static EventDate build(final String date, final Event event) {
-		Objects.requireNonNull(event, "event cannot be null");
-		final String protoDate = checkDate(Util.checkString(date, "date"));
-		return new EventDate(protoDate, event);
-	}
-
-	/** Ensures that a string can be parsed as a valid date.throws
+	 * Ensures that a string can be parsed as a valid date.
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/EventDate.java:68]: Moved the `parseDate` function above the `build` functions so the object constructors are at the bottom of the file.

File: src/us/kbase/workspace/database/provenance/EventDate.java [hunk 2]
```diff
@@ -107,14 +83,78 @@ private static String checkDate(final String protoDate) {
...
-			}
-			catch (DateTimeParseException e) {
+			} catch (DateTimeParseException e) {
 				// report the error below as an IllegalArgumentException
 			}
 		}
-		throw new IllegalArgumentException("Invalid date: \"" + protoDate + "\"\ndate must be in the format yyyy, yyyy-MM, or yyyy-MM-dd and be a valid combination of day, month, and year.");
+		throw new IllegalArgumentException(
+				"Invalid date: \"" + protoDate
+						+ "\"\ndate must be in the format yyyy, yyyy-MM, or yyyy-MM-dd and be a valid combination of day, month, and year.");
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/EventDate.java:96]: horrible choices by the auto-formatter

File: src/us/kbase/workspace/database/provenance/EventDate.java [hunk 3]
```diff
@@ -107,14 +83,78 @@ private static String checkDate(final String protoDate) {
...
+			protoDate = checkDate(trimmedDate);
+		} catch (Exception e) {
+			errorList.add(e.getMessage());
+		}
+
+		if (errorList.isEmpty()) {
+			return new EventDate(protoDate, event);
+		}
+		throw new IllegalArgumentException(
+				"Errors in EventDate construction:\n" + String.join("\n", errorList));
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/EventDate.java:158]: all construction errors are collected and thrown as as `IllegalArgumentException`

File: src/us/kbase/workspace/test/database/provenance/EventDateTest.java [hunk 4]
```diff
@@ -122,27 +121,43 @@ public void buildAndTrimEvent() throws Exception {
...
+	@Test
+	public void buildFailNullOrWhitespaceEventDate() throws Exception {
+		for (String nullOrWs : WHITESPACE_STRINGS_WITH_NULL) {
+			buildEventDateFailWithError(nullOrWs, nullOrWs, EVENT_NO_NULL_WS + "\n" + DATE_NO_NULL_WS);
+		}
+	}
 
-			buildEventDateFailWithError(nullOrWs, nullOrWs, new IllegalArgumentException("event cannot be null or whitespace only"));
+	@Test
+	public void buildEventDateFailDateEvent() throws Exception {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/EventDateTest.java:155]: reporting more than one error in the attempt to construct the EventDate object. Brave new world!

**Review by MrCreosote — COMMENTED** [review]
I don't see a problem with the changes, but I do have a question about error propagation. Presumably these are going to be embedded in another credit class in a list or map or something. How do you plan to propagate these errors to the user? The current error would make it difficult to debug the issue if there are several dozen of these in a list

**Comment by ialarmedalien** [comment]
All the errors are going to be "rolled up" to the top level container.

**Review by MrCreosote — APPROVED** [review]
Ok, I assume that'll also provide info as to the location of the error in the data structure

**Pull request merged**